docs/usage/routing/index.rst

```diff
@@ -1,9 +1,412 @@
+
+.. literalinclude:: /examples/routing/dynamic_registration.py
+    :language: python
+
+
+.. attention::
+    It is generally advised to only use this pattern if strictly necessary. In most
+    cases it is possible to work around this by using plugins or the factory pattern.
+
+
+Reusing handlers and controllers under different paths
+++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+Route handlers and controllers can be registered multiple times under different paths:
+
+.. tab-set::
+
+    .. tab-item:: Handler
+
+        .. literalinclude:: /examples/routing/reusing_1.py
+            :language: python
+
+    .. tab-item:: Controller
+
+        .. literalinclude:: /examples/routing/reusing_2.py
+            :language: python
+
+
+Defining paths
+--------------
+
+Every component can define its path, that is, the part of the HTTP request's URL the
+component should respond to. Paths are organized hierarchically, with every component
+using the path of its ancestor as its base path. In practice this means that a route
+handler with the path ``/handler`` registered on a router with the path ``/router``
+will be respond to the path ``/router/handler``.
+
+Handlers
+++++++++
+
+On a route handler, that path can be set using the ``path`` argument:
+
+.. tab-set::
+
+    .. tab-item:: Positional
+
+        .. literalinclude:: /examples/routing/handler_path_positional.py
+            :language: python
+
+    .. tab-item:: Keyword
+
+        .. literalinclude:: /examples/routing/handler_path_kwarg.py
+            :language: python
+
+
+When not provided, the path defaults to ``/``.
+
+
+Controllers
++++++++++++
+
+The path on a controller can be set with the ``path`` attribute:
+
+.. literalinclude:: /examples/routing/controller_path.py
+    :language: python
+
+
+When not provided, the path defaults to ``/``.
+
+Routers
++++++++
+
+The path on a router can be set via the ``path`` argument. Unlike handlers and
+controllers, specifying a path is not optional.
+
+.. literalinclude:: /examples/routing/router_path.py
+    :language: python
+
+
+Setting a base path for the application
+++++++++++++++++++++++++++++++++++++++++
+
+Sometimes it is desirable to set a base path for the entire application. While it is
+generally advisable to handle this outside the application itself (e.g. via a reverse
+proxy), this can be achieved by registering all route handlers with a router, and
+setting the base path there.
+
+
+Dynamic paths with path parameters
+++++++++++++++++++++++++++++++++++
+
+Dynamic paths can be created with the help of path parameters. These are part of the
+path string and follow the basic form of ``{<parameter name>:<parameter type>}``, where
+``<parameter name>`` is the name of the parameter used to make its value available to
+the route handler and what will be represented in the OpenAPI schema, and
+``<parameter type>`` a basic Python type that this parameter will be validated against
+and, if possible, converted to.
```

Comment:
> These are part of the
path string and follow the basic form of ``{<parameter name>:<parameter type>}``, where
``<parameter name>`` is the name of the parameter used to make its value available to
the route handler and what will be represented in the OpenAPI schema, and
``<parameter type>`` a basic Python type that this parameter will be validated against
and, if possible, converted to.

This is quite a big sentence - might be better to break it up into smaller ideas, e.g., a sentence that introduces the two components of the parameter, then one explaining parameter name, and another to explain parameter type.